### PR TITLE
Add 0dim Tensor overload for _foreach_div

### PR DESCRIPTION
**Pull request by janeyx99**
This PR is ALMOST basically just following the steps from #106677 EXCEPT! We do add one feature. Similar to fused_adam(w), for the CUDA dispatches: when the scalar tensor is on CPU, we .item and redispatch to the normal scalar overload. Otherwise, the cuda kernel will complain about mismatch in devices between the scalar and the tensors.

Why do we add this feature? Our optimizers want to allow lr as a tensor, and lr could be a CPU tensor. lr is used with foreach_div_ in Adam, so our CI will break otherwise.

After this PR, `_foreach_mul` and `_foreach_div` will accept either a CPU or a GPU tensor for the scalar tensor (vs only a GPU tensor). They join the ranks of `fused_adam(w)` in this characteristic. I did not yet do the same thing for foreach_add (the only other foreach op with a .Tensor overload) because there is no use case and will be more involved.

cc @crcrpar 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113688

### CODE REVIEW

File: test/test_foreach.py [hunk 1]
```diff
@@ -48,7 +48,7 @@ def __init__(self, func):
         # Some foreach functions don't have in-place implementations.
         self.is_inplace = False if func is None else func.__name__.endswith('_')
 
-    def __call__(self, inputs, is_cuda, is_fastpath, **kwargs):
+    def __call__(self, inputs, is_cuda, expect_fastpath, **kwargs):
```
Comment (janeyx99) [inline test/test_foreach.py:51]: rename to expect_fastpath, which is what it really is and is less confusing

**Review by mlazos — APPROVED** [review]

File: test/test_foreach.py [hunk 2]
```diff
@@ -772,12 +774,24 @@ def test_tensors_grouping(self):
...
+        self.assertEqual(actual, [t.div(scalar_cpu_tensor) for t in tensors])
+
+
     @onlyCUDA
     def test_0dim_tensor_overload_exception(self):
         # check exceptions of fast path
         tensors = [make_tensor((2, 2), dtype=torch.float, device="cuda") for _ in range(2)]
         with self.assertRaisesRegex(RuntimeError, "scalar tensor expected to be on"):
-            torch._foreach_mul(tensors, torch.tensor(1.0, device="cpu"))
+            torch._foreach_add(tensors, torch.tensor(1.0, device="cpu"), alpha=1.0)
```
Comment (janeyx99) [inline test/test_foreach.py:794]: This works now, but made me realize I didn't add a case for _foreach_add when I added the overload. Adding that now.

**Review by albanD — APPROVED** [review]
Please confirm there is a test case that also ensures that a list of cpu Tensor and a list of cuda Tensor properly raises.

**Comment by janeyx99** [comment]
> Please confirm there is a test case that also ensures that a list of cpu Tensor and a list of cuda Tensor properly raises.

test_parity tests that both a forloop over the ops + the foreach op will return the same thing or error the same:
https://github.com/pytorch/pytorch/blob/42b2b9e663a085f5a4ef0f61b5e11f4756521400/test/test_foreach.py#L163-L175

Note that if the reference code also errors in line 174, the test will fail.

**Comment by mlazos** [comment]
> Please confirm there is a test case that also ensures that a list of cpu Tensor and a list of cuda Tensor properly raises.

@janeyx99 hmm at what level will this error, I probably need to add the grouping at the dynamo level if this occurs when FakeTensor tracing as we discussed

**Comment by janeyx99** [comment]
> @janeyx99 hmm at what level will this error, I probably need to add the grouping at the dynamo level if this occurs when FakeTensor tracing as we discussed

Ah, what happens is that we'll dispatch into the CUDA impl `foreach_tensor_div_tensor_kernel_cuda` which is a C++ function that checks device types. Since the device types do not match, we then dispatch to the slow impl, which is just another C++ function that for loops over the normal `div`. Once it calls into `div`, it will error. It's worth ensuring what torch.compile will do (hopefully it's not too smart; we don't want it converting the second list to GPU tensors).

**Comment by mlazos** [comment]
> > @janeyx99 hmm at what level will this error, I probably need to add the grouping at the dynamo level if this occurs when FakeTensor tracing as we discussed
> 
> Ah, what happens is that we'll dispatch into the CUDA impl `foreach_tensor_div_tensor_kernel_cuda` which is a C++ function that checks device types. Since the device types do not match, we then dispatch to the slow impl, which is just another C++ function that for loops over the normal `div`. Once it calls into `div`, it will error. It's worth ensuring what torch.compile will do (hopefully it's not too smart; we don't want it converting the second list to GPU tensors).

Nah this is good, this is exactly what I want. I also will only apply the special handling we discussed when there is a single tensor in the second arg.

**Pull request closed without merging**